### PR TITLE
ci: Fix build not including typescript types

### PR DESCRIPTION
**Pull request by brandonlenz**
# Summary

Looks like when the TS version was bumped, the refactor to remove `awesome-typescript-loader` and rely purely on babel didn't backfilll for type generation which babel does not do on its own. In this PR we manually call `tsc` in the build step and let it emit the type declarations. This is the recommended approach for projects using babel alone and not relying on ts-loader.

Alternative considered: use ts loader on ts|x files before running babel. ts-loader is a bit slow, and felt like a more repetitive solution, using two loaders on all of our ts/tsx files when `tsc` will directly create just the type exports for us without any additional config.

Resources: 
- https://www.typescriptlang.org/docs/handbook/babel-with-typescript.html

## Related Issues or PRs

closes: https://github.com/trussworks/react-uswds/issues/2053
stems from: https://github.com/trussworks/react-uswds/pull/1548

## How To Test

Checkout the code, clean your local files (delete node_modules and lib), freshly install packages with yarn. running `yarn` will also run `yarn build` which will create the lib directory where you can confirm that .d.ts files are now output correctly again. Running `yarn pack` will generate the tarball which you can also verify includes the types.

```
> git checkout bl-fix-types
> rm -rf node_modules
> rm -rf lib
> yarn
> yarn pack
```

### Screenshots (optional):
image

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -93,11 +98,12 @@ export { Logo } from './components/Footer/Logo/Logo'
 export { SocialLinks } from './components/Footer/SocialLinks/SocialLinks'
 
 /** Modal components */
-export { Modal, ModalProps, ModalRef } from './components/Modal/Modal'
+export { Modal } from './components/Modal/Modal'
 export { ModalToggleButton } from './components/Modal/ModalToggleButton'
 export { ModalOpenLink } from './components/Modal/ModalOpenLink'
 export { ModalHeading } from './components/Modal/ModalHeading/ModalHeading'
 export { ModalFooter } from './components/Modal/ModalFooter/ModalFooter'
+export type { ModalProps, ModalRef } from './components/Modal/Modal'
```
Comment (brandonlenz) [inline src/index.ts:106]: These `export type` changes silence some TS warnings we were getting on build

File: tsconfig.json [hunk 2]
```diff
@@ -11,11 +11,12 @@
     "module": "esnext",
     "moduleResolution": "node",
     "resolveJsonModule": true,
-    "noEmit": true,
+    "emitDeclarationOnly": true,
```
Comment (brandonlenz) [inline tsconfig.json:14]: Previously, `awesome-typescript-loader` was exporting the types. Now we rely on tsc to do it, so we emit those type declarations only, and let babel emit the rest of the transpiled source code

File: tsconfig.json [hunk 3]
```diff
@@ -11,11 +11,12 @@
...
     "jsx": "react",
     "noImplicitAny": true,
     "declaration": true,
-    "outDir": "./lib"
+    "outDir": "./lib",
+    "isolatedModules": true,
```
Comment (brandonlenz) [inline tsconfig.json:19]: Recommended setting from https://www.typescriptlang.org/docs/handbook/babel-with-typescript.html

https://www.typescriptlang.org/tsconfig#isolatedModules

File: package.json [hunk 4]
```diff
@@ -27,9 +27,9 @@
     "test:serverside": "yarn build && node src/serverSideTest.js",
     "storybook": "start-storybook -p 9009",
     "storybook:deploy": "storybook-to-ghpages",
-    "build": "webpack --progress",
-    "build:watch": "webpack --watch",
-    "lint": "tsc --noEmit && eslint --ext js,jsx,ts,tsx src && stylelint \"src/**/*.{css,scss}\"",
+    "build": "tsc && webpack --progress",
+    "build:watch": "tsc && webpack --watch",
```
Comment (brandonlenz) [inline package.json:31]: Manually call `tsc` here so that types are built to `lib`

File: package.json [hunk 5]
```diff
@@ -27,9 +27,9 @@
...
+    "lint": "tsc --noEmit --emitDeclarationOnly false && eslint --ext js,jsx,ts,tsx src && stylelint \"src/**/*.{css,scss}\"",
```
Comment (brandonlenz) [inline package.json:32]: `noEmit` cannot be used alongside `emitDeclarationOnly` (as of ts 3.8), which is now in tsconfig. So we have to set it to false here, which is the recommended workaround

**Review by Igarfinkle — APPROVED** [review]
This tentatively looks good to me... how can we check that this fixes the problem?

**Comment by brandonlenz** [comment]
> This tentatively looks good to me... how can we check that this fixes the problem?

The `How to test` steps involve running the commands that our release CI runs to package the library. Following those steps results in properly packaged type definitions.

**Pull request merged**